Added test coverage for ProductService, ProductCategoryService, integrations and api

Agarwalvidu opened the pull request:
This PR includes all the rquirements till Week 5 along with unit tests and integration tests. Please review and suggest improvements.

Agarwalvidu committed:
Added unit and integration tests for ProductService and ProductCategoryService

Agarwalvidu committed:
Updated Week 6

tavyandy97 commented on frontend/package.json:
```diff
@@ -10,8 +10,8 @@
     "@types/react-dom": "^18.0.0",
-    "react": "^19.0.0",
-    "react-dom": "^19.0.0",
+    "react": "^18.3.1",
+    "react-dom": "^18.3.1",
```
Any specific reason we've downgraded to react@18.x?

Agarwalvidu replied:
I downgraded to React 18.3.1 because I encountered compatibility issues with some dependency, I am unable to recall it. But I am actively testing with React 19 again, will upgrade.

tavyandy97 replied:
Sure thing, lemme know what the issue is whenever you run into it 👍🏻

tavyandy97 commented on frontend/product.js:
```diff
@@ -0,0 +1,26 @@
+        tile.innerHTML = `
+            <div class="product-title">${product.name}</div>
+            <div class="product-price">₹${product.price}</div>
+            <p>${product.description || ""}</p>
+            <p class="product-brand">${product.brand || ""}</p>
```
nitpick: Both of these tags can be conditionally added depending upon if the value of the variable inside them exists

Agarwalvidu replied:
Thank you for pointing that out! I have updated the code to conditionally render the tags only when there is content. I’ will definitely keep this in mind going forward.

tavyandy97 commented on frontend/product.js:
```diff
@@ -0,0 +1,26 @@
+            <p>${product.description || ""}</p>
+            <p class="product-brand">${product.brand || ""}</p>
+          `;
+        container.appendChild(tile);
```
Recommendation:
Instead of appending elements one by one, we can use `DocumentFragment` (`document.createDocumentFragment`), which will help us avoid incrementally altering the DOM.

Repainting the DOM is a relatively heavier operation and would be better if we did it only once.

Agarwalvidu replied:
Thanks a lot for the recommendation!
I was not aware of DocumentFragment before, this was a great opportunity to learn about it. I have refactored the code to use a fragment to optimize DOM updates and will definitely keep such performance considerations in mind for future implementations as well.

tavyandy97 commented on frontend/product.js:
```diff
@@ -0,0 +1,26 @@
+  fetch("http://127.0.0.1:8000/api/mongo/products/")
+    .then((res) => res.json())
+    .then((data) => {
+      console.log("Fetched products:", data);
```
Maybe we can add some validation checks on data? There might be some cases where the data might not have come or the API call might've failed.

Agarwalvidu replied:
That totally makes sense, I have now added checks to handle unexpected or empty responses(like I have added a check if the length of products data is 0 then it would display "No products found." and other checks if there is some issue with the API call). Also kept the technical logs in the console for debugging.

tavyandy97 replied:
A check needs to be placed on the status as well for non 2xx HTTP status codes

tavyandy97 commented on backend/tests/seed_data.py:
```diff
@@ -0,0 +1,42 @@
+from products_mongo.models import Product, ProductCategory
+from products_mongo.test_connection import init_db
+
+init_db
```
Shouldn't this be a function call?

Agarwalvidu replied:
Thank you for catching it, I missed the parenthesis. I have corrected it to init_db() so the DB initialization now runs properly.

tavyandy97 commented on backend/tests/seed_data.py:
```diff
@@ -0,0 +1,42 @@
+        quantity_in_warehouse=200,
+        category=category2
+    )
+    product2.save()
```
When writing integration tests, it's a better practice to add multiple entries just to simulate a better DB state

Agarwalvidu replied:
I will surely add more product entries. Will keep this in mind for future test setups as well!

tavyandy97 commented:
For the `backend/tests/test_product_category_services.py` file, we don't have any test cases for invalid inputs or error cases. It'd be a good practice to add those as well

tavyandy97 commented on frontend/product.css:
```diff
@@ -0,0 +1,49 @@
+    transform: scale(1.05);
+  }
+  
+  .product-title {
```
The CSS for this selector is written two times.

Agarwalvidu replied:
Thanks for pointing that out! I have cleaned up the duplicate CSS for .product-title.

tavyandy97 commented on frontend/product.css:
```diff
@@ -0,0 +1,49 @@
+body {
```
Recommendation: 
Can consider using CSS variables like 
```
:root {
  --color-background: #f8f8f8;
  --color-price: #2e8b57;
  --color-shadow: rgba(0, 0, 0, 0.1);
}
```

Agarwalvidu replied:
Thank you for the suggestion! I have implemented the CSS variables to make the code more maintainable.

tavyandy97 commented:
A couple of cases that would be good to cover in `backend/tests/test_product_services.py`:
- Validation tests for:
    - Price negative values
    - Invalid date ranges
    - Duplicate product names
    - Invalid category assignments
- Making a separate file for mock data

Agarwalvidu commented:
I have ve now added test cases for missing fields, invalid IDs, and failure cases in update/delete operations in backend/tests/test_product_category_services.py. Will make sure to include such edge cases going forward too.
I have added validation tests for price negative values, invalid date ranges, invalid category and duplicate product names in backend/tests/test_product_services.py. I have seperate file for mock data as well.
Thank you pointing these things, I learned a lot while adding these validations and I will keep these in mind as well.

Agarwalvidu committed:
Changes made according to the suggestions by the mentor

Agarwalvidu committed:
Merge branch 'code-review' of https://github.com/Agarwalvidu/interneers-lab into code-review

Agarwalvidu commented:
@tavyandy97 Please review the recent changes! Thank you.

tavyandy97 commented:
Seems like there's a redundant set of frontend files. Removing the old set of files would be a good idea 😄

tavyandy97 commented on backend/django_app/settings.py:
```diff
@@ -15,6 +15,9 @@
 
 # Build paths inside the project like this: BASE_DIR / 'subdir'.
 BASE_DIR = Path(__file__).resolve().parent.parent
```
Seems like some duplicate lines have been added in this file

tavyandy97 commented on backend/tests/test_product_category_services.py:
```diff
@@ -0,0 +1,88 @@
+
+    @patch('products_mongo.product_category_repository.ProductCategoryRepository.create_category')
+    def test_create_category_with_invalid_input(self, mock_create_category):
+        """Test category creation with missing fields"""
+        mock_create_category.side_effect = ValueError("Title is required")
+
+        with self.assertRaises(ValueError):
+            ProductCategoryService.create_category({"description": "No title provided"})
```
- If the `ValueError` is raised by the service due to input validation before calling the repository, then mocking the repository is irrelevant - it will never be called.
- If the repository is responsible for raising the ValueError, then mocking it makes sense. Now it's just testing the mocked behavior, not real logic.

tavyandy97 replied:
Same comment can be applied for other tests which have been added
